File: data_structures/trees/binary_search_tree/python/binary_search_tree_create_from_sorted.py
from typing import List, TypeVar, Union
from data_structures.trees.binary_tree.python.binary_tree_node import BinaryTreeNode

T = TypeVar('T')


def sorted_array_to_bst_recursive(array: List[T]) -> Union[None, BinaryTreeNode[T]]:
    if not array or len(array) <= 0:
        return None

    mid = len(array) // 2

    root = BinaryTreeNode(array[mid])

    # Note: slicing array is expensive. See optimal recursive solution below
    root.left = sorted_array_to_bst_recursive(array[:mid])
    root.right = sorted_array_to_bst_recursive(array[mid + 1:])

    return root


def sorted_array_to_bst_recursive_optimized(array: List[T]) -> Union[None, BinaryTreeNode[T]]:
    return sorted_array_to_bst_recursive_optimal_build(array, 0, len(array))


def sorted_array_to_bst_recursive_optimal_build(array: List[T], low: int, high: int) -> Union[None, BinaryTreeNode[T]]:
    if low == high:
        return None

    mid = (low + high) // 2
    node = BinaryTreeNode(array[mid])
    node.left = sorted_array_to_bst_recursive_optimal_build(array, low, mid)
    node.right = sorted_array_to_bst_recursive_optimal_build(array, mid + 1, high)

    return node


if __name__ == "__main__":
    import doctest

    doctest.testmod(verbose=True)

File: data_structures/trees/binary_search_tree/typescript/binarySearchTreeCreateFromSorted.ts
import BinaryTreeNode from '@ds/trees/binary_tree/typescript/BinaryTreeNode';

export function sortedArrayToBstRecursive<T>(array: T[]): BinaryTreeNode<T> {
    if (!array || array.length === 0) {
        return null;
    }

    const mid = Math.floor(array.length / 2);
    const root = new BinaryTreeNode(array[mid]);

    root.left = sortedArrayToBstRecursive(array.slice(0, mid));
    root.right = sortedArrayToBstRecursive(array.slice(mid + 1));

    return root;
}

export function sortedArrayToBstRecursiveOptimized<T>(array: T[]): BinaryTreeNode<T> {
    if (!array || array.length === 0) {
        return null;
    }

    return sortedArrayToBstRecursiveOptimizedBuild(array, 0, array.length - 1);
}

function sortedArrayToBstRecursiveOptimizedBuild<T>(array: T[], low: number, high: number): BinaryTreeNode<T> {
    if (low === high) {
        return null;
    }

    const mid = Math.floor((low + high) / 2);
    const root = new BinaryTreeNode(array[mid]);

    root.left = sortedArrayToBstRecursiveOptimizedBuild(array, low, mid);
    root.right = sortedArrayToBstRecursiveOptimizedBuild(array, mid + 1, high);

    return root;
}
